### ISSUE TITLE
Whether the hashCode of the class object will conflict？

### ISSUE THREAD

**Issue by qcks** [dart-lang/sdk#44512]
flutter does not support reflection, I think hascode may meet my needs,I wonder if the hashCode of the class object will conflict?

**Comment by lrhn** [comment]
Different objects may have the same hash code. The only thing that hash code promises is that *equal* objects have the *same* hash code. (Technically, `int get hashCode => 4;` would be a *valid* implementation, just a very inefficient one).
The `hashCode` of `Object` is the same as `identitiyHashCode`. It promises only that the same object gives the same result each time, not that two different objects have different identity hash codes.

**Comment by qcks** [comment]
`int _getHashCode<T>({T t}) {
  return T.hashCode;
}

void main() {
  print("\t Obj1.hashCode =${new Object().hashCode}");
  print("\t Obj2.hashCode =${new Object().hashCode}");

  print("\t Object.hashCode =${(Object).hashCode}");
  print("\t Object.hashCode =${(Object).hashCode}");

  print("\t List<Object>.hashCode =${_getHashCode<List<Object>>()}");
  print("\t List<Object>.hashCode =${_getHashCode<List<Object>>()}");

  print("\t 1null.hashCode =${_getHashCode(t: null)}");
  print("\t 2null.hashCode =${_getHashCode<Null>()}");
}`

**Comment by srawlins** [comment]
This is  a language question if `Type.hashCode` is specified in the spec. It is a VM and other implementations question otherwise.

**Comment by tatumizer** [comment]
@qcks: could you describe the practical problem are you trying to solve? Maybe it can be solved without relying on the uniqueness of hashCode?

**Comment by qcks** [comment]
There may be infinitely many instance objects, but the class template is finite. there can be (Class template).hashCode in Dart, and if I can use it, then I don't need to use additional parameters and judgments to control my project.

My problem: parsing json in flutter. my project framework started coupling this json generation plugin [flutterjsonbeanfactory](https://plugins.jetbrains.com/plugin/11415-flutterjsonbeanfactory ), but he has problems: 1. can not use obfuscation; 2. using generic toString, T.toString will conflict (a.dart{ class User} == b.dart{ class User} ).
My solution: no longer rely strongly on other json generation plug-ins, interface-oriented (including fromJson, toJson methods can be), giving business developers freedom within the module; do not rely on the incoming generic<T> to match the generated bean (which requires (Class template) of the unique property), but use the generic non-empty instance object.

**Comment by qcks** [comment]
> This is a language question if `Type.hashCode` is specified in the spec. It is a VM and other implementations question otherwise.

There may be infinitely many instance objects, but the class template is finite. there can be (Class template).hashCode in Dart.This is unique to Dart, I don't know what he can do.Thank you.

**Comment by tatumizer** [comment]
@qcks : Sorry, I don't understand your explanation. Can you formulate the problem in a form that doesn't involve the concepts like class templates, bean factories etc.  From all your prior posts, I could gather that you are trying to create a mapping between types and something else (I don't know what it is, but it doesn't matter). In which case, you can use type object as a whole as a key, e.g.
```
void main() {
  var s = [List, String, int, Object, Object];
  var h = <Type, dynamic>{};
  s.forEach((element) {
    h.putIfAbsent(element, () {
      print("\t add ${element.toString()}");
      return "some value associated with $element";
    });
  });
  print("$h");
}
```
It prints:
```
{List<dynamic>: some value associated with List<dynamic>, String: some value associated with String, int: some value associated with int, Object: some value associated with Object}
```
Maybe this is what you are looking for?

**Comment by eernstg** [comment]
@qcks, I think we can close this issue: As @tatumizer mentioned, the apparent use case allows you to use the `Type` objects directly rather than their hash code. The language specification specifies very few methods, and in particular it is up to the system libraries (like `'dart:core'` or `'dart:async'`) to define and document the behavior of methods like `hashCode`; so this would be a system library issue rather than a language issue.

@qcks, if there are any further ideas to explore here, or problems to consider, then please create a new issue!

**Comment by qcks** [comment]
Thanks @eernstg , sorry I'm late, this is how I solved the problem at the time [
```
getT<T>({T? t}) => T;
///  if( T is List<User>)//Earlier versions of darts reported errors
if (T == getT<void>() || T == getT<Map<User,User2>() || T == getT<List<User>>()) {
  .......
}
```

]. Nowadays a high percentage of flutter apps still make heavy use of T.toString==‘xxx’ which is a security risk in the long run.

**Comment by eernstg** [comment]
Right, `T.toString() == 'someString'` is not safe and should never be used. For instance, class names can be changed by minimization.
